vertical-pod-autoscaler/pkg/admission-controller/logic/recommendation_provider.go

```diff
@@ -31,104 +29,63 @@ import (
-		return nil, false
-	}
-	// originalLimit set but originalRequest not set - K8s will treat the pod as if they were equal,
-	// recommend limit equal to request
-	if originalRequest == nil || originalRequest.Value() == 0 {
-		result := *recommendedRequest
-		return &result, false
-	}
-	// originalLimit and originalRequest are set. If they are equal recommend limit equal to request.
-	if originalRequest.MilliValue() == originalLimit.MilliValue() {
-		result := *recommendedRequest
-		return &result, false
-	}
-
-	// Input and output milli values should fit in int64 but intermediate values might be bigger.
-	originalMilliRequest := big.NewInt(originalRequest.MilliValue())
-	originalMilliLimit := big.NewInt(originalLimit.MilliValue())
-	recommendedMilliRequest := big.NewInt(recommendedRequest.MilliValue())
-	var recommendedMilliLimit big.Int
-	recommendedMilliLimit.Mul(recommendedMilliRequest, originalMilliLimit)
-	recommendedMilliLimit.Div(&recommendedMilliLimit, originalMilliRequest)
-	if recommendedMilliLimit.IsInt64() {
-		return resource.NewMilliQuantity(recommendedMilliLimit.Int64(), recommendedRequest.Format), false
-	}
-	return resource.NewMilliQuantity(math.MaxInt64, recommendedRequest.Format), true
-}
-
 // GetContainersResources returns the recommended resources for each container in the given pod in the same order they are specified in the pod.Spec.
-func GetContainersResources(pod *v1.Pod, podRecommendation vpa_types.RecommendedPodResources, annotations vpa_api_util.ContainerToAnnotationsMap) []ContainerResources {
-	resources := make([]ContainerResources, len(pod.Spec.Containers))
+func GetContainersResources(pod *v1.Pod, podRecommendation vpa_types.RecommendedPodResources, limitRange *v1.LimitRangeItem,
+	annotations vpa_api_util.ContainerToAnnotationsMap) []vpa_api_util.ContainerResources {
+	resources := make([]vpa_api_util.ContainerResources, len(pod.Spec.Containers))
+	var defaultCpu, defaultMem, maxCpuLimit, maxMemLimit *resource.Quantity
+	if limitRange != nil {
+		defaultCpu = limitRange.Default.Cpu()
+		defaultMem = limitRange.Default.Memory()
+		maxCpuLimit = limitRange.Max.Cpu()
+		maxMemLimit = limitRange.Max.Memory()
+	}
 	for i, container := range pod.Spec.Containers {
-		resources[i] = newContainerResources()
-
 		recommendation := vpa_api_util.GetRecommendationForContainer(container.Name, &podRecommendation)
 		if recommendation == nil {
 			klog.V(2).Infof("no matching recommendation found for container %s", container.Name)
 			continue
 		}
-		resources[i].Requests = recommendation.Target
-
-		cpuLimit, capped := getProportionalLimit(container.Resources.Limits.Cpu(), container.Resources.Requests.Cpu(), resources[i].Requests.Cpu())
-		if cpuLimit != nil {
-			resources[i].Limits[v1.ResourceCPU] = *cpuLimit
-		}
+		cpuLimit, capped := vpa_api_util.GetProportionalLimit(container.Resources.Limits.Cpu(), container.Resources.Requests.Cpu(), recommendation.Target.Cpu(), defaultCpu)
```

Comment:
Can we have GetProportionalLimit to return annotation instead of bool in case of capping? I think it would make this code clearer.

Reply:
Done.